payment_methods/SquareOnsite/forms/SettingsForm.php

```diff
@@ -94,186 +94,114 @@ public function addSquareConnectButton(EE_PMT_SquareOnsite $paymentMethod, EE_Pa
-            $pmInstance->delete_extra_meta(Domain::META_KEY_APPLICATION_ID);
-            $pmInstance->delete_extra_meta(Domain::META_KEY_ACCESS_TOKEN);
-            $pmInstance->delete_extra_meta(Domain::META_KEY_LOCATION_ID);
-            $pmInstance->update_extra_meta(
-                Domain::META_KEY_SQUARE_DATA,
-                [
-                    Domain::META_KEY_USING_OAUTH => false,
-                ]
+        // If there is an established connection we should check the debug mode and the connection.
+        if (
+            isset($this->squareData[ Domain::META_KEY_LIVE_MODE ])
+            && $this->squareData[ Domain::META_KEY_LIVE_MODE ]
+            && $pmDebugMode
+        ) {
+            EED_SquareOnsiteOAuth::errorLogAndExit($pmInstance, 'Debug vs live mode', $this->squareData, false);
+            $this->add_validation_error(
+                sprintf(
+                    // translators: %1$s: opening strong html tag. $2$s: closing strong html tag.
+                    esc_html__(
+                        '%1$sSquare Payment Method%2$s is in debug mode but the authentication with %1$sSquare%2$s is in Live mode. Payments will not be processed correctly! If you wish to test this payment method, please reset the connection and use sandbox credentials to authenticate with Square.',
+                        'event_espresso'
+                    ),
+                    '<strong>',
+                    '</strong>'
+                ),
+                'ee4_square_live_connection_but_pm_debug_mode'
+            );
+        } elseif (
+            (! isset($this->squareData[ Domain::META_KEY_LIVE_MODE ])
+            || ! $this->squareData[ Domain::META_KEY_LIVE_MODE ])
+            && ! $pmDebugMode
+        ) {
+            EED_SquareOnsiteOAuth::errorLogAndExit($pmInstance, 'Debug vs live mode', $this->squareData, false);
+            $this->add_validation_error(
+                sprintf(
+                    // translators: %1$s: opening strong html tag. $2$s: closing strong html tag.
+                    esc_html__(
+                        '%1$sSquare Payment Method%2$s is in live mode but the authentication with %1$sSquare%2$s is in sandbox mode. Payments will not be processed correctly! If you wish to process real payments with this payment method, please reset the connection and use live credentials to authenticate with Square.',
+                        'event_espresso'
+                    ),
+                    '<strong>',
+                    '</strong>'
+                ),
+                'ee4_square_sandbox_connection_but_pm_not_in_debug_mode'
             );
-        } catch (EE_Error $e) {
-            return false;
         }
-        return true;
+        // Don't allow changing the debug mode setting while connected.
+        // And if we're creating the form for receiving POST data, ignore debug mode input's value.
+        if (method_exists($debugInput, 'isDisabled')) {
+            $debugInput->disable();
```

Comment:
uhhh... we're checking for a method named `isDisabled` and then calling `disable()` ? 
Is that correct? Usually, the called method would be the same one checked, ie:

```suggestion
        if (method_exists($debugInput, 'disable')) {
            $debugInput->disable();
```

or

```suggestion
        if (method_exists($debugInput, 'isDisabled')) {
            $debugInput->isDisabled();
```

if we have control over whatever the class the `$debugInput` object is, we could also create an interface like:

```php
interface Disableable 
{
    public function isDisabled();

    public function disable();
}
```

to indicate that that object has that API (literally HAS that interface) and then check it like:

```php
if ($debugInput instanceof Disableable) {
```

Reply:
🤔 🤔

Reply:
don't worry about the interface idea, that's just more food for thought

but the method exists check followed by a call to a different method is concerning and should maybe be looked into
if it's correct as it, then maybe a code comment would help explain why it is the way it is